File: backend/app/schemas/entities.py
"""ARCH42-S1:schemas — entity graph API shapes. Mirrored field for field by
frontend/src/types/entities.ts (verify_arch42 gate F2)."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field


class EntityPotential(BaseModel):
    documents: int


class ModelRow(BaseModel):
    kind: str
    version: Optional[int] = None
    source: str
    pair_count: int = 0
    converged: bool = False
    fitted_at: Optional[datetime] = None


class EntitySummary(BaseModel):
    counts_by_kind: dict[str, int]
    records: int
    documents_linked: int
    open_reviews: int
    open_conflicts: int
    models: list[ModelRow]


class EntityRow(BaseModel):
    id: uuid.UUID
    kind: str
    display_name: str
    documents: int
    mentions: int
    identifier_kinds: list[str]
    merged_records: int
    last_seen_at: Optional[datetime] = None


class EntityList(BaseModel):
    items: list[EntityRow]
    total: int
    page: int
    page_size: int
    matched_by_identifier: bool = False


class IdentifierRow(BaseModel):
    id: uuid.UUID
    kind: str
    display: str
    derived: bool
    entity_id: uuid.UUID


class DocumentRow(BaseModel):
    mention_id: uuid.UUID
    work_item_id: uuid.UUID
    filename: str
    role: str
    field_path: str
    decision: str
    method: str
    probability: Optional[float] = None
    entity_id: uuid.UUID
    created_at: datetime


class RelationshipRow(BaseModel):
    relation: str
    direction: str
    other_id: uuid.UUID
    other_kind: str
    other_name: str
    documents: int


class MemberRow(BaseModel):
    id: uuid.UUID
    display_name: str
    merged_at: Optional[datetime] = None
    merge_reason: Optional[str] = None
    mentions: int


class CandidateRow(BaseModel):
    id: uuid.UUID
    left_id: uuid.UUID
    left_name: str
    right_id: uuid.UUID
    right_name: str
    kind: str
    reason: str
    probability: float
    weight: float
    conflict_kinds: list[str]
    status: str
    comparison: dict[str, Optional[int]] = Field(default_factory=dict)
    created_at: datetime


class ObligationsPlaceholder(BaseModel):
    available: bool = False
    milestone: str
    items: list[dict] = Field(default_factory=list)


class Entity360(BaseModel):
    entity: EntityRow
    first_seen_at: Optional[datetime] = None
    split_from_id: Optional[uuid.UUID] = None
    identifiers: list[IdentifierRow]
    documents: list[DocumentRow]
    relationships: list[RelationshipRow]
    members: list[MemberRow]
    open_candidates: list[CandidateRow]
    obligations: ObligationsPlaceholder


class GraphNode(BaseModel):
    id: uuid.UUID
    kind: str
    label: str
    documents: int
    depth: int


class GraphEdge(BaseModel):
    source: uuid.UUID
    target: uuid.UUID
    relation: str
    weight: int


class EntityGraph(BaseModel):
    root_id: uuid.UUID
    nodes: list[GraphNode]
    edges: list[GraphEdge]
    truncated: bool


class MergeRequest(BaseModel):
    into_entity_id: uuid.UUID


class SplitRequest(BaseModel):
    mention_ids: list[uuid.UUID] = Field(min_length=1, max_length=500)


class RenameRequest(BaseModel):
    display_name: str = Field(min_length=1, max_length=300)


class EntityChip(BaseModel):
    entity_id: uuid.UUID
    kind: str
    display_name: str
    role: str
    decision: str
    field_path: str


class WorkItemEntities(BaseModel):
    chips: list[EntityChip]
    resolved: bool


class ResolveResult(BaseModel):
    resolved: bool
    detail: dict[str, int | bool | str] = Field(default_factory=dict)

File: frontend/src/types/entities.ts
/** ARCH42-S2:types — mirrors backend/app/schemas/entities.py field for field (verify_arch42 E10). */

export type EntityKind = "PERSON" | "ORGANIZATION" | "ADDRESS" | "ACCOUNT" | "ASSET" | "SHIPMENT";
export const ENTITY_KINDS: readonly EntityKind[] = ["PERSON", "ORGANIZATION", "ADDRESS", "ACCOUNT", "ASSET", "SHIPMENT"];
export const ENTITY_KIND_LABELS: Readonly<Record<EntityKind, string>> = {
  PERSON: "People",
  ORGANIZATION: "Organizations",
  ADDRESS: "Addresses",
  ACCOUNT: "Accounts",
  ASSET: "Assets",
  SHIPMENT: "Shipments",
};

export interface EntityPotential {
  readonly documents: number;
}

export interface ModelRow {
  readonly kind: string;
  readonly version?: number | null;
  readonly source: string;
  readonly pair_count: number;
  readonly converged: boolean;
  readonly fitted_at?: string | null;
}

export interface EntitySummary {
  readonly counts_by_kind: Readonly<Record<string, number>>;
  readonly records: number;
  readonly documents_linked: number;
  readonly open_reviews: number;
  readonly open_conflicts: number;
  readonly models: readonly ModelRow[];
}

export interface EntityRow {
  readonly id: string;
  readonly kind: EntityKind;
  readonly display_name: string;
  readonly documents: number;
  readonly mentions: number;
  readonly identifier_kinds: readonly string[];
  readonly merged_records: number;
  readonly last_seen_at?: string | null;
}

export interface EntityList {
  readonly items: readonly EntityRow[];
  readonly total: number;
  readonly page: number;
  readonly page_size: number;
  readonly matched_by_identifier: boolean;
}

export interface IdentifierRow {
  readonly id: string;
  readonly kind: string;
  readonly display: string;
  readonly derived: boolean;
  readonly entity_id: string;
}

export interface DocumentRow {
  readonly mention_id: string;
  readonly work_item_id: string;
  readonly filename: string;
  readonly role: string;
  readonly field_path: string;
  readonly decision: string;
  readonly method: string;
  readonly probability?: number | null;
  readonly entity_id: string;
  readonly created_at: string;
}

export interface RelationshipRow {
  readonly relation: string;
  readonly direction: string;
  readonly other_id: string;
  readonly other_kind: string;
  readonly other_name: string;
  readonly documents: number;
}

export interface MemberRow {
  readonly id: string;
  readonly display_name: string;
  readonly merged_at?: string | null;
  readonly merge_reason?: string | null;
  readonly mentions: number;
}

export interface CandidateRow {
  readonly id: string;
  readonly left_id: string;
  readonly left_name: string;
  readonly right_id: string;
  readonly right_name: string;
  readonly kind: string;
  readonly reason: string;
  readonly probability: number;
  readonly weight: number;
  readonly conflict_kinds: readonly string[];
  readonly status: string;
  readonly comparison: Readonly<Record<string, number | null>>;
  readonly created_at: string;
}

export interface ObligationsPlaceholder {
  readonly available: boolean;
  readonly milestone: string;
  readonly items: readonly unknown[];
}

export interface Entity360 {
  readonly entity: EntityRow;
  readonly first_seen_at?: string | null;
  readonly split_from_id?: string | null;
  readonly identifiers: readonly IdentifierRow[];
  readonly documents: readonly DocumentRow[];
  readonly relationships: readonly RelationshipRow[];
  readonly members: readonly MemberRow[];
  readonly open_candidates: readonly CandidateRow[];
  readonly obligations: ObligationsPlaceholder;
}

export interface GraphNode {
  readonly id: string;
  readonly kind: EntityKind;
  readonly label: string;
  readonly documents: number;
  readonly depth: number;
}

export interface GraphEdge {
  readonly source: string;
  readonly target: string;
  readonly relation: string;
  readonly weight: number;
}

export interface EntityGraph {
  readonly root_id: string;
  readonly nodes: readonly GraphNode[];
  readonly edges: readonly GraphEdge[];
  readonly truncated: boolean;
}

export interface EntityChip {
  readonly entity_id: string;
  readonly kind: EntityKind;
  readonly display_name: string;
  readonly role: string;
  readonly decision: string;
  readonly field_path: string;
}

export interface WorkItemEntities {
  readonly chips: readonly EntityChip[];
  readonly resolved: boolean;
}

export interface ResolveResult {
  readonly resolved: boolean;
  readonly detail: Readonly<Record<string, number | boolean | string>>;
}

export const RELATION_LABELS: Readonly<Record<string, string>> = {
  EMPLOYED_BY: "employed by",
  CONTRACTED_WITH: "contracted with",
  SUPPLIES: "supplies",
  LEASES_FROM: "leases from",
  OCCUPIES: "occupies",
  LESSOR_OF: "lessor of",
  SHIPPED: "shipped",
  CONSIGNED_TO: "consigned to",
  CONTAINS: "contains",
  HOLDS_ACCOUNT: "holds account",
  LOCATED_AT: "located at",
};
